File: tests/test_drake_smart_ptr.py
# -*- coding: utf-8 -*-
import pytest
import sys
import weakref

import env  # noqa: F401

from pybind11_tests import drake_smart_ptr as m
from pybind11_tests import ConstructorStats


def test_pointer_caster():
    assert m.test_pointer_caster()


@pytest.mark.skip(
    reason="Generally reproducible in CPython, Python 3, non-debug, on Linux. "
    "However, hard to pin this down for CI."
)
def test_1922():
    # Test #1922 (drake#11424).
    # Define a derived class which *does not* overload the method.
    # WARNING: The reproduction of this failure may be platform-specific, and
    # seems to depend on the order of definition and/or the name of the classes
    # defined. For example, trying to place this and the C++ code in
    # `test_virtual_functions` makes `assert id_1 == id_2` below fail.
    class Child1(m.ExampleVirt2):
        pass

    id_1 = id(Child1)
    assert m.example_virt2_get_name(m.ExampleVirt2()) == "ExampleVirt2"
    assert m.example_virt2_get_name(Child1()) == "ExampleVirt2"

    # Now delete everything (and ensure it's deleted).
    wref = weakref.ref(Child1)
    del Child1
    pytest.gc_collect()
    assert wref() is None

    # Define a derived class which *does* define an overload.
    class Child2(m.ExampleVirt2):
        def get_name(self):
            return "Child2"

    id_2 = id(Child2)
    assert id_1 == id_2  # This happens in CPython; not sure about PyPy.
    assert m.example_virt2_get_name(m.ExampleVirt2()) == "ExampleVirt2"
    # THIS WILL FAIL: This is using the cached `ExampleVirt2.get_name`, rather
    # than re-inspect the Python dictionary.
    assert m.example_virt2_get_name(Child2()) == "Child2"


@pytest.mark.skipif(env.PYPY, reason="Unsupported on PyPy")
def test_mi_ownership_constraint():
    # See below for positive tests.

    # unique_ptr
    c = m.ContainerBase1(m.MIType(10, 100))
    assert c is not None

    # shared_ptr
    # Should not throw an error.
    obj = m.Base12a(10, 100)
    assert obj.bar() == 100
    c = m.ContainerBase2a(obj)
    # Use variable to avoid style violations.
    assert c is not None

    # TODO(eric.cousineau): This currently causes a segfault in both
    # this branch and on `master` (a303c6f).
    # Figure out if there is a way to fix this?

    # assert c.get().bar() == 100

    # # Should throw an error.
    # c = m.ContainerBase2a(m.Bae12a(10, 100))
    # assert c.get().foo() == 10
    # assert c.get().bar() == 100


def define_child(name, base_type, stats_type):
    # Derived instance of `DefineBase<>` in C++.
    # `stats_type` is meant to enable us to use `ConstructorStats` exclusively for a Python class.

    class ChildT(base_type):
        def __init__(self, *args):
            base_type.__init__(self, *args)
            self.icstats = m.get_instance_cstats(ChildT.get_cstats(), self)
            self.icstats.track_created()

        def __del__(self):
            self.icstats.track_destroyed()

        def value(self):
            # Use a different value, so that we can detect slicing.
            return 10 * base_type.value(self)

        @staticmethod
        def get_cstats():
            return ConstructorStats.get(stats_type)

    ChildT.__name__ = name
    return ChildT


ChildBad = define_child("ChildBad", m.BaseBad, m.ChildBadStats)
Child = define_child("Child", m.Base, m.ChildStats)

ChildBadUnique = define_child("ChildBadUnique", m.BaseBadUnique, m.ChildBadUniqueStats)
ChildUnique = define_child("ChildUnique", m.BaseUnique, m.ChildUniqueStats)


# TODO(eric.cousineau): See if this is at all possibly on PyPy.
# Placing `pytest.gc_collect` near `del` statements indicates that we are not
# capturing deletion properly.
@pytest.mark.skipif(env.PYPY, reason="Unsupported on PyPy")
def test_shared_ptr_derived_slicing(capture):
    leaked_count = [0]
    is_py38 = sys.version_info[:2] >= (3, 8)

    def py38_leak():
        if is_py38:
            leaked_count[0] += 1

    def cstats_alive_except_leaked():
        return cstats.alive() - leaked_count[0]

    # [ Bad ]
    cstats = ChildBad.get_cstats()
    # Create instance in move container to permit releasing.
    obj = ChildBad(10)
    obj_weak = weakref.ref(obj)
    # This will release the reference, the refcount will drop to zero, and Python will destroy it.
    c = m.BaseBadContainer(obj)
    del obj
    # We will have lost the derived Python instance.
    assert obj_weak() is None
    # Check stats:
    assert cstats.alive() == 0
    # As an additional check, we will try to query the value from the container's value.
    # This should have been 100 if the trampoline had retained its Python portion.
    assert c.get().value() == 10
    # Destroy references.
    del c

    # [ Good ]
    # See above for setup.
    cstats = Child.get_cstats()
    # Try a temporary setup.
    c = m.BaseContainer(Child(10))
    assert cstats.alive() == 1
    obj = c.release()
    del c
    assert cstats.alive() == 1
    assert obj.value() == 100
    del obj
    py38_leak()
    assert cstats_alive_except_leaked() == 0
    # Use something more permanent.
    obj = Child()  # Use factory method.
    obj_weak = weakref.ref(obj)
    assert cstats_alive_except_leaked() == 1
    c = m.BaseContainer(obj)
    del obj
    assert cstats_alive_except_leaked() == 1
    # We now still have a reference to the object. py::wrapper<> will intercept Python's
    # attempt to destroy `obj`, is aware the `shared_ptr<>.use_count() > 1`, and will increase
    # the ref count by transferring a new reference to `py::wrapper<>` (thus reviving the object,
    # per Python's documentation of __del__).
    assert obj_weak() is not None
    assert cstats_alive_except_leaked() == 1
    # This goes from C++ -> Python, and then Python -> C++ once this statement has finished.
    assert c.get().value() == 100
    assert cstats_alive_except_leaked() == 1
    # Destroy references (effectively in C++), and ensure that we have the desired behavior.
    del c
    py38_leak()
    assert cstats_alive_except_leaked() == 0

    # Ensure that we can pass it from Python -> C++ -> Python, and ensure that C++ does not think
    # that it has ownership.
    obj = Child(10)
    c = m.BaseContainer(obj)
    del obj
    assert cstats_alive_except_leaked() == 1
    obj = c.get()
    # Now that we have it in Python, there should only be 1 Python reference, since
    # py::wrapper<> in C++ should have released its reference.
    expected_refcount = 2
    if is_py38:
        expected_refcount += 1
    assert sys.getrefcount(obj) == expected_refcount
    del c
    assert cstats_alive_except_leaked() == 1
    assert obj.value() == 100
    del obj
    py38_leak()
    assert cstats_alive_except_leaked() == 0


@pytest.mark.skipif(env.PYPY, reason="Unsupported on PyPy")
def test_unique_ptr_derived_slicing(capture):
    # [ Bad ]
    cstats = ChildBadUnique.get_cstats()
    # Create instance in move container to permit releasing.
    try:
        c = m.BaseBadUniqueContainer(ChildBadUnique(10))
    except RuntimeError as e:
        assert "wrapper<>" in str(e)
    # We lost this portion to slicing.
    assert cstats.alive() == 0

    # [ Good ]
    cstats = ChildUnique.get_cstats()
    # Try a temporary setup.
    c = m.BaseUniqueContainer(ChildUnique(10))
    assert cstats.alive() == 1
    obj = c.release()
    del c
    assert cstats.alive() == 1
    assert obj.value() == 100
    del obj
    assert cstats.alive() == 0

    # Ensure that we can pass between Python -> C++ -> Python
    # (using factory method).
    obj = m.BaseUniqueContainer(ChildUnique()).release()
    obj = m.BaseUniqueContainer(obj).release()
    assert obj.value() == 100
    del obj
    assert cstats.alive() == 0


def test_unique_ptr_arg():
    stats = ConstructorStats.get(m.UniquePtrHeld)

    pass_through_list = [
        m.unique_ptr_pass_through,
        m.unique_ptr_pass_through_move_to_py,
        m.unique_ptr_pass_through_cast_to_py,
        # TODO(eric.cousineau): Fix these cases.
        # m.unique_ptr_pass_through_cast_from_py,
        # m.unique_ptr_pass_through_move_from_py,
    ]
    for pass_through in pass_through_list:
        obj = m.UniquePtrHeld(1)
        obj_ref = pass_through(obj)
        assert stats.alive() == 1
        assert obj.value() == 1
        assert obj == obj_ref
        del obj
        del obj_ref
        pytest.gc_collect()
        assert stats.alive() == 0

    obj = m.UniquePtrHeld(1)
    m.unique_ptr_terminal(obj)
    assert stats.alive() == 0

    m.unique_ptr_terminal(m.UniquePtrHeld(2))
    assert stats.alive() == 0

    assert m.unique_ptr_pass_through(None) is None
    m.unique_ptr_terminal(None)

    with pytest.raises(TypeError):
        m.unique_ptr_terminal(m.UniquePtrOther())


def test_unique_ptr_to_shared_ptr():
    obj = m.shared_ptr_held_in_unique_ptr()
    assert m.shared_ptr_held_func(obj)


def test_unique_ptr_overload_fail():
    obj = m.UniquePtrHeld(1)
    # These overloads pass ownership back to Python.
    out = m.unique_ptr_overload(obj, m.FirstT())
    assert out["obj"] is obj
    assert out["overload"] == 1
    out = m.unique_ptr_overload(obj, m.SecondT())
    assert out["obj"] is obj
    assert out["overload"] == 2


def test_unique_ptr_held_container_from_cpp():
    def check_reset(obj_new):
        c = m.UniquePtrHeldContainer()
        obj = c.get()
        assert obj is not None
        obj_ref = c.reset(obj_new)
        assert obj is obj_ref
        assert c.get() is obj_new

    check_reset(obj_new=m.UniquePtrHeld(10))
    check_reset(obj_new=None)

File: tests/test_drake_smart_ptr.cpp
/*
    tests/test_ownership_transfer.cpp -- test ownership transfer semantics.

    Copyright (c) 2017 Eric Cousineau <eric.cousineau@tri.global>

    All rights reserved. Use of this source code is governed by a
    BSD-style license that can be found in the LICENSE file.
*/

#if defined(_MSC_VER) && _MSC_VER < 1910
#  pragma warning(disable: 4702) // unreachable code in system header
#endif

#include <memory>

#include "pybind11_tests.h"
#include "object.h"

enum Label : int {
  BaseBadLabel,
  ChildBadLabel,
  BaseLabel,
  ChildLabel,

  BaseBadUniqueLabel,
  ChildBadUniqueLabel,
  BaseUniqueLabel,
  ChildUniqueLabel,
};

// For attaching instances of `ConstructorStats`.
template <int label>
class Stats {};

struct Base1 {
    Base1(int i) : i(i) { }
    int foo() { return i; }
    int i;
};

struct Base2a {
    Base2a(int i) : i(i) { }
    int bar() { return i; }
    int i;
};

template <typename T, typename Ptr = std::unique_ptr<T>>
class Container {
public:
    Container(Ptr ptr)
        : ptr_(std::move(ptr)) {}
    T* get() const { return ptr_.get(); }
    Ptr release() { return std::move(ptr_); }

    static void def(py::module &m, const std::string& name) {
        py::class_<Container>(m, name.c_str())
            .def(py::init<Ptr>())
            .def("get", &Container::get)
            .def("release", &Container::release);
    }
private:
    Ptr ptr_;
};

template <int label>
class DefineBase {
 public:
  DefineBase(int value)
      : value_(value) {
    track_created(this, value);
  }
  // clang does not like having an implicit copy constructor when the
  // class is virtual (and rightly so).
  DefineBase(const DefineBase&) = delete;
  virtual ~DefineBase() {
    track_destroyed(this);
  }
  virtual int value() const { return value_; }
 private:
  int value_{};
};

template <int label>
class DefineBaseContainer {
 public:
  using T = DefineBase<label>;
  DefineBaseContainer(std::shared_ptr<T> obj)
      : obj_(obj) {}
  std::shared_ptr<T> get() const { return obj_; }
  std::shared_ptr<T> release() { return std::move(obj_); }
 private:
  std::shared_ptr<T> obj_;
};

template <int label>
class DefineBaseUniqueContainer {
 public:
  using T = DefineBase<label>;
  DefineBaseUniqueContainer(std::unique_ptr<T> obj)
      : obj_(std::move(obj)) {}
  T* get() const { return obj_.get(); }
  std::unique_ptr<T> release() { return std::move(obj_); }
 private:
  std::unique_ptr<T> obj_;
};

template <int label>
class DefinePyBase : public DefineBase<label> {
 public:
  using BaseT = DefineBase<label>;
  using BaseT::BaseT;
  int value() const override {
    PYBIND11_OVERLOAD(int, BaseT, value);
  }
};

template <int label>
class DefinePyBaseWrapped : public py::wrapper<DefineBase<label>> {
 public:
  using BaseT = py::wrapper<DefineBase<label>>;
  using BaseT::BaseT;
  int value() const override {
    PYBIND11_OVERLOAD(int, BaseT, value);
  }
};

// BaseBad - No wrapper alias.
using BaseBad = DefineBase<BaseBadLabel>;
using BaseBadContainer = DefineBaseContainer<BaseBadLabel>;
using ChildBadStats = Stats<ChildBadLabel>;

// Base - wrapper alias used in pybind definition.
using Base = DefineBase<BaseLabel>;
using PyBase = DefinePyBase<BaseLabel>;
using BaseContainer = DefineBaseContainer<BaseLabel>;
using ChildStats = Stats<ChildLabel>;

// - Unique Ptr
// BaseBad - No wrapper alias.
using BaseBadUnique = DefineBase<BaseBadUniqueLabel>;
using BaseBadUniqueContainer = DefineBaseUniqueContainer<BaseBadUniqueLabel>;
using ChildBadUniqueStats = Stats<ChildBadUniqueLabel>;

// Base - wrapper alias used directly.
using BaseUnique = DefineBase<BaseUniqueLabel>;
using PyBaseUnique = DefinePyBaseWrapped<BaseUniqueLabel>;
using BaseUniqueContainer = DefineBaseUniqueContainer<BaseUniqueLabel>;
using ChildUniqueStats = Stats<ChildUniqueLabel>;

class PyInstanceStats {
 public:
  PyInstanceStats(ConstructorStats& cstats, py::handle h)
    : cstats_(cstats),
      h_(h) {}
  void track_created() {
    cstats_.created(h_.ptr());
    cstats_.value(py::str(h_).cast<std::string>());
  }
  void track_destroyed() {
    cstats_.destroyed(h_.ptr());
  }
 private:
  ConstructorStats& cstats_;
  py::handle h_;
};

PyInstanceStats get_instance_cstats(ConstructorStats& cstats, py::handle h) {
  return PyInstanceStats(cstats, h);
}

template <typename C, typename... Args>
using class_shared_ = py::class_<C, Args..., std::shared_ptr<C>>;

template <typename... Args>
using class_unique_ = py::class_<Args...>;

class UniquePtrHeld {
public:
    UniquePtrHeld() = delete;
    UniquePtrHeld(const UniquePtrHeld&) = delete;
    UniquePtrHeld(UniquePtrHeld&&) = delete;

    UniquePtrHeld(int value)
        : value_(value) {
        print_created(this, value);
    }
    ~UniquePtrHeld() {
        print_destroyed(this);
    }
    int value() const { return value_; }
private:
    int value_{};
};

TEST_SUBMODULE(drake_smart_ptr, m) {
  // For Drake issue: https://github.com/RobotLocomotion/drake/issues/9398
  m.def("test_pointer_caster", []() -> bool {
      UserType a;
      UserType *a_ptr = &a;
      py::object o = py::cast(&a); // Rvalue
      py::object o1 = py::cast(a_ptr); // Non-rvalue
      return (py::cast<UserType*>(o) == a_ptr && py::cast<UserType*>(o1) == a_ptr);
  });

  Container<Base1>::def(m, "ContainerBase1");

  Container<Base2a, std::shared_ptr<Base2a>>::def(m, "ContainerBase2a");

  // Test #1922 (drake#11424).
  class ExampleVirt2 {
      public:
          virtual ~ExampleVirt2() = default;
          virtual std::string get_name() const { return "ExampleVirt2"; }
      };
  class PyExampleVirt2 : public ExampleVirt2 {
  public:
      std::string get_name() const override {
          PYBIND11_OVERLOAD(std::string, ExampleVirt2, get_name, );
      }
  };
  py::class_<ExampleVirt2, PyExampleVirt2>(m, "ExampleVirt2")
      .def(py::init())
      .def("get_name", &ExampleVirt2::get_name);
  m.def("example_virt2_get_name",
      [](const ExampleVirt2& obj) { return obj.get_name(); });

  // No alias - will not have lifetime extended.
  class_shared_<BaseBad>(m, "BaseBad")
      .def(py::init<int>())
      .def("value", &BaseBad::value);
  class_shared_<BaseBadContainer>(m, "BaseBadContainer")
      .def(py::init<std::shared_ptr<BaseBad>>())
      .def("get", &BaseBadContainer::get)
      .def("release", &BaseBadContainer::release);
  class_shared_<ChildBadStats>(m, "ChildBadStats");

  // Has alias - will have lifetime extended.
  class_shared_<Base, py::wrapper<PyBase>>(m, "Base")
      .def(py::init<int>())
      // Factory method for alias.
      .def(py::init([]() { return new py::wrapper<PyBase>(10); }))
      .def("value", &Base::value);
  class_shared_<BaseContainer>(m, "BaseContainer")
      .def(py::init<std::shared_ptr<Base>>())
      .def("get", &BaseContainer::get)
      .def("release", &BaseContainer::release);
  class_shared_<ChildStats>(m, "ChildStats");

  class_unique_<BaseBadUnique>(m, "BaseBadUnique")
      .def(py::init<int>())
      .def("value", &BaseBadUnique::value);
  class_unique_<BaseBadUniqueContainer>(m, "BaseBadUniqueContainer")
      .def(py::init<std::unique_ptr<BaseBadUnique>>())
      .def("get", &BaseBadUniqueContainer::get)
      .def("release", &BaseBadUniqueContainer::release);
  class_unique_<ChildBadUniqueStats>(m, "ChildBadUniqueStats");

  class_unique_<BaseUnique, PyBaseUnique>(m, "BaseUnique")
      .def(py::init<int>())
      // Factory method.
      .def(py::init([]() { return new PyBaseUnique(10); }))
      .def("value", &BaseUnique::value);
  class_unique_<BaseUniqueContainer>(m, "BaseUniqueContainer")
      .def(py::init<std::unique_ptr<BaseUnique>>())
      .def("get", &BaseUniqueContainer::get)
      .def("release", &BaseUniqueContainer::release);
  class_unique_<ChildUniqueStats>(m, "ChildUniqueStats");

  class_shared_<PyInstanceStats>(m, "InstanceStats")
      .def(py::init<ConstructorStats&, py::handle>())
      .def("track_created", &PyInstanceStats::track_created)
      .def("track_destroyed", &PyInstanceStats::track_destroyed);
  m.def("get_instance_cstats", &get_instance_cstats);


    py::class_<UniquePtrHeld>(m, "UniquePtrHeld")
        .def(py::init<int>())
        .def("value", &UniquePtrHeld::value);

    class UniquePtrOther {};
    py::class_<UniquePtrOther>(m, "UniquePtrOther")
        .def(py::init<>());

    m.def("unique_ptr_pass_through",
        [](std::unique_ptr<UniquePtrHeld> obj) {
            return obj;
        });
    m.def("unique_ptr_terminal",
        [](std::unique_ptr<UniquePtrHeld> obj) {
            obj.reset();
            return nullptr;
        });

    // Check traits in a concise manner.
    static_assert(
        py::detail::move_common<std::unique_ptr<UniquePtrHeld>>::value,
        "This must be true.");

    // Guarantee API works as expected.
    m.def("unique_ptr_pass_through_cast_from_py",
        [](py::object obj_py) {
            auto obj =
                py::cast<std::unique_ptr<UniquePtrHeld>>(std::move(obj_py));
            return obj;
        });
    m.def("unique_ptr_pass_through_move_from_py",
        [](py::object obj_py) {
            return py::move<std::unique_ptr<UniquePtrHeld>>(std::move(obj_py));
        });

    m.def("unique_ptr_pass_through_move_to_py",
        [](std::unique_ptr<UniquePtrHeld> obj) {
            return py::move(std::move(obj));
        });

    m.def("unique_ptr_pass_through_cast_to_py",
        [](std::unique_ptr<UniquePtrHeld> obj) {
            return py::cast(std::move(obj));
        });

    class FirstT {};
    py::class_<FirstT>(m, "FirstT")
        .def(py::init());
    class SecondT {};
    py::class_<SecondT>(m, "SecondT")
        .def(py::init());

    m.def("unique_ptr_overload",
        [](std::unique_ptr<UniquePtrHeld> obj, FirstT) {
            py::dict out;
            out["obj"] = py::cast(std::move(obj));
            out["overload"] = 1;
            return out;
        });
    m.def("unique_ptr_overload",
        [](std::unique_ptr<UniquePtrHeld> obj, SecondT) {
            py::dict out;
            out["obj"] = py::cast(std::move(obj));
            out["overload"] = 2;
            return out;
        });

    // Ensure class is non-empty, so it's easier to detect double-free
    // corruption. (If empty, this may be harder to see easily.)
    struct SharedPtrHeld { int value = 10; };
    py::class_<SharedPtrHeld, std::shared_ptr<SharedPtrHeld>>(m, "SharedPtrHeld")
        .def(py::init<>());
    m.def("shared_ptr_held_in_unique_ptr",
        []() {
            return std::unique_ptr<SharedPtrHeld>(new SharedPtrHeld());
        });
    m.def("shared_ptr_held_func",
        [](std::shared_ptr<SharedPtrHeld> obj) {
            return obj != nullptr && obj->value == 10;
        });

    // Test passing ownership of registered, but unowned, C++ instances back to
    // Python. This happens when a raw pointer is passed first, and then
    // ownership is transfered.
    struct UniquePtrHeldContainer {
        UniquePtrHeldContainer() {
            value_.reset(new UniquePtrHeld(10));
        }
        UniquePtrHeld* get() const {
            return value_.get();
        }
        using Ptr = std::unique_ptr<UniquePtrHeld>;
        Ptr reset(Ptr to) {
            Ptr from = std::move(value_);
            value_ = std::move(to);
            return from;
        }
        std::unique_ptr<UniquePtrHeld> value_;
    };
    py::class_<UniquePtrHeldContainer>(m, "UniquePtrHeldContainer")
        .def(py::init())
        .def("get", &UniquePtrHeldContainer::get, py::return_value_policy::reference_internal)
        .def("reset", &UniquePtrHeldContainer::reset);
}
